### PR TITLE
[specific ci=3-03-Docker-Compose-Basic] Unbind container in containerRm

### LINKED ISSUE #5948: docker-compose integration tests not cleaning up on failures
Seen in build https://ci.vcna.io/vmware/vic/12992:

screen shot 2017-08-07 at 23 14 21

The `Compose Up with link` test failed before getting to the `docker-compose down` step, thereby leaving orphan containers connected to the `vic_default` network, which caused subsequent `docker-compose down` commands to fail with

```
...
Removing network vic_default
vic_default has active endpoints
```

**Proposed solutions:**

* docker-compose integration test suites use a `Test Teardown  Run Keyword If Test Failed` step to clean up orphan containers if a test fails before the cleanup step (`docker-compose down`).
* See if there's a docker-compose command option to make orphan containers attached to networks non-fatal for `docker-compose down`.

### PR DESCRIPTION
**Pull request by chengwang86**
Fixes #5948 

This PR tries to unbind the container in the before removing the container. 

This may potentially degrade the performance of containerRm. However, since the PL will not do anything if the container is already unbound, its impact on performance may be neligible.

### CODE REVIEW

**Comment by hmahmood** [comment]
Will this actually solve the issue with `kill` not actually unbinding the container from the network?

**Comment by chengwang86** [comment]
@hmahmood I think the original issue can be fixed by doing this. The reason why the failure occurs is that the container is removed before it is unbound, and that `compose down` calls `container stop` followed by `container rm ` instead of `container rm -f`. So this PR unbinds the container in the code path where the containerRM() is called without `-f`.

**Comment by chengwang86** [comment]
Deleting `dangling containerVMs` from network layer when calling `deleteScope` is also an option, but it requires that the network layer knows about the status of the exec cache, which violates the isolation of layers.

**Comment by hmahmood** [comment]
So if call `docker kill` for a container, it can potentially still be connected to its network afterwards if I don't follow it with an `rm`?

**Comment by hmahmood** [comment]
Why do we not have a special case for `SIGKILL` in https://github.com/vmware/vic/blob/master/lib/apiservers/engine/backends/container.go#L634 as we do on the portlayer side (https://github.com/vmware/vic/blob/master/lib/portlayer/exec/container.go#L454), and unbind the container there?

**Comment by chengwang86** [comment]
@hmahmood If `docker kill` is not followed by `docker rm`, the container will not be removed from cache and the containerPoweredOff event will finally be processed properly. So the container will be unbound from the network in this case.

The failure occurs because `docker kill` is followed by `docker rm` before the event is processed.

About why I didn't have a special case checking `SIGKILL` in `containerKill()`, George's opinion was that we should target at a more general set of failures that might occur due to this race condition, instead of only focusing on `docker kill`. @hickeng Thoughts?

**Comment by hmahmood** [comment]
What about port mappings? Do they also get unmapped when with a `kill`?

**Comment by chengwang86** [comment]
@hmahmood ContainerKill() calls container_proxy.Signal(), which would **unmap** the ports if the state is not `running`. So I think the ports should've been unmapped after `docker kill`.

In this case, I think I can move `unmapPorts()` outside of `UnbindContainerfFromNetwork()`.

File: lib/apiservers/engine/backends/container_proxy.go [hunk 1]
```diff
@@ -835,6 +826,28 @@ func (c *ContainerProxy) Stop(vc *viccontainer.VicContainer, name string, second
...
+		switch err := err.(type) {
+		case *scopes.UnbindContainerNotFound:
+			// ignore error
+			log.Warnf("Container %s not found by network unbind", vc.ContainerID)
+		case *scopes.UnbindContainerInternalServerError:
+			return "", InternalServerError(err.Payload.Message)
+		default:
+			return "", InternalServerError(err.Error())
+		}
+	} else {
```
Comment (anchal-agrawal) [inline lib/apiservers/engine/backends/container_proxy.go]: Don't think the `else` block is needed - we can just say

```go
    if {
        ...
    }

    return ub.Payload.Handle, nil
```

File: lib/apiservers/engine/backends/container_proxy.go [hunk 2]
```diff
@@ -835,6 +826,28 @@ func (c *ContainerProxy) Stop(vc *viccontainer.VicContainer, name string, second
 	return nil
 }
 
+func (c *ContainerProxy) UnbindContainerFromNetwork(vc *viccontainer.VicContainer, handle string) (string, error) {
```
Comment (anchal-agrawal) [inline lib/apiservers/engine/backends/container_proxy.go]: We should have function comments for exported functions.

File: lib/apiservers/engine/backends/container.go [hunk 3]
```diff
@@ -761,6 +761,16 @@ func (c *Container) ContainerRm(name string, config *types.ContainerRmConfig) er
 		if state.Status == "Starting" {
 			return derr.NewRequestConflictError(fmt.Errorf("The container is starting.  To remove use -f"))
 		}
+
+		handle, err := c.Handle(id, name)
+		if err != nil {
+			return err
```
Comment (anchal-agrawal) [inline lib/apiservers/engine/backends/container.go]: Is this a docker-typed error?

Comment (chengwang86) [reply lib/apiservers/engine/backends/container.go]: I think so. It returns either `NotFoundError` or `InternalServerError`.

File: lib/apiservers/engine/backends/container_proxy.go [hunk 4]
```diff
@@ -835,6 +826,27 @@ func (c *ContainerProxy) Stop(vc *viccontainer.VicContainer, name string, second
 	return nil
 }
 
+// UnbindContainerFromNetwork unbinds a container from the networks that it connects to
+func (c *ContainerProxy) UnbindContainerFromNetwork(vc *viccontainer.VicContainer, handle string) (string, error) {
+	defer trace.End(trace.Begin(""))
```
Comment (jzt) [inline lib/apiservers/engine/backends/container_proxy.go]: Would it be worthwhile to trace on the container ID here?

**Pull request merged**